### PR TITLE
Refactored badge overlay

### PR DESCRIPTION
**Pull request by maxxcrawford**
Revised badge overlay to be psuedo/class based to injected HTML element.

### CODE REVIEW

File: src/content_script.js [hunk 1]
```diff
@@ -32,6 +33,100 @@ function itemWidthCheck (target) {
...
+
+  // Check for Badge element and select it
+  if (!badgeClassUId) {
+    badgeClassUId = "js-" + target;
+  }
+
+  htmlBadgeDiv = document.querySelector("." + badgeClassUId);
+
+  // Confirm target element is defined
+  if ( target && typeof target === "object" ) {
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply src/content_script.js]: 👍

**Comment by maxxcrawford** [comment]
@groovecoder Ready for re-review!

File: src/content_script.js [hunk 2]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
+}
+
+function closePrompt() {
+  const activePrompt = findActivePrompt();
+  activePrompt.classList.remove("active");
+  document.body.classList.remove("js-fbc-prompt-active");
+  document.querySelector(".fbc-has-badge.js-fbc-prompt-active").classList.remove("js-fbc-prompt-active");
+}
+
+function positionPrompt ( target ) {
```
Comment (groovecoder) [inline src/content_script.js]: Change this param to something besides `target` - e.g., `targetSelector` (and drop the logic to prepend the `.` character). Better to not change the type (and meaning) of a function parameter inside the function.

File: src/content_script.js [hunk 3]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
+    htmlBadgeDiv.classList.remove("fbc-badge-fixed");
+  } else if ( isFixed(target) ) {
+    htmlBadgeDiv.classList.add("fbc-badge-fixed");
+    return {offsetPosX: elemRect.left, offsetPosY: elemRect.top};
+  } else {
+    return {offsetPosX: elemRect.left - bodyRect.left, offsetPosY: elemRect.top - bodyRect.top};
+  }
+}
+
+function checkVisibilityAndAppleClass(target, htmlBadgeDiv) {
```
Comment (groovecoder) [inline src/content_script.js]: `appleclass` ? 🍎 for the class?

File: src/content_script.js [hunk 4]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
+  // Check for Badge element and select it
+  if (!badgeClassUId) {
+    badgeClassUId = "js-" + target;
+  }
+
+  const htmlBadgeDiv = document.querySelector("." + badgeClassUId);
+
+  // Confirm target element is defined
+  if (!target || !(typeof target === "object")) {
+    target = document.querySelector("." + target);
```
Comment (groovecoder) [inline src/content_script.js]: This changes `target` variable type from a string to an element? Let's keep the type the same. So maybe the parameter is `targetClass` or `targetUId` and then this is `const targetElement`

File: src/content_script.js [hunk 5]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
+  }
+
+  checkVisibilityAndAppleClass(target, htmlBadgeDiv);
```
Comment (groovecoder) [inline src/content_script.js]: 🍎Class

File: src/content_script.js [hunk 6]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
+
+  if (typeof smallSwitch === "undefined") {
+    if (htmlBadgeDiv.classList.contains("fbc-badge-small")) {
+      smallSwitch = true;
+    }
+  }
+
+  // Set offset size based on large/small badge
+  const [elementSizeOffsetX, elementSizeOffsetY] = elementSizeOffsetXY(smallSwitch);
```
Comment (groovecoder) [inline src/content_script.js]: Array de-structuring assignments ... so beautiful. They should have sent a poet.

File: src/content_script.js [hunk 7]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
+        const itemUIDClassName = "fbc-UID_" + (facebookDetectedElementsArr.length + 1);
+        const itemUIDClassTarget = "js-" + itemUIDClassName;
+        const socialAction = "login";
+        facebookDetectedElementsArr.push(itemUIDClassName);
+        addFacebookBadge(item, itemUIDClassTarget, socialAction);
+        item.classList.add("fbc-has-badge");
+        item.classList.add(itemUIDClassName);
+      }
+    }
+  }
```
Comment (groovecoder) [inline src/content_script.js]: Lots of code duplication and nested `for` loops here. Should be refactored into a single function that can accept `socialAction` as parameter.

File: src/content_script.js [hunk 8]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
-  }, 10);
+    screenUpdate();
+  }
+});
+
+let checkForTrackers = true;
+
+browser.runtime.onMessage.addListener(message => {
+  console.log("browser.runtime.onMessage");
+  console.log(message["msg"]);
```
Comment (groovecoder) [inline src/content_script.js]: Will need to drop these before we release.

File: src/content_script.js (hunk 3 above)
Comment (maxxcrawford) [reply src/content_script.js]: 😂 Typo casualty.

File: src/_locales/en/messages.json [hunk 9]
```diff
@@ -1,142 +1,166 @@
...
+	"onboarding2-alt": {
+		"message": "The Facebook Container fence icon",
+		"description": "This is alternative text for the icon in the second onboarding panel"
+	},
+	"onboarding2-p1": {
+		"message": "This icon appears in the Firefox address bar when you visit sites, like Facebook.com, that are in the Facebook Container boundary.",
+		"description": "This is the first paragraph of the second onboarding panel."
+	},
+	"onboarding2-p2": {
+		"message": "These sites need to access to your Facebook identity (or know who you are) in order to work. *SPANSTART This also means that Facebook can track everything you do on these sites. *SPANEND",
```
Comment (stoically) [inline src/_locales/en/messages.json]: Just a small typo; I believe it should be just "to access" instead of "to access to".

Comment (maxxcrawford) [reply src/_locales/en/messages.json]: Revised!

File: src/_locales/en/messages.json [hunk 10]
```diff
@@ -1,142 +1,166 @@
...
+		"description": ""
+	},
+	"inPageUI-tooltip-prompt-p2": {
+		"message": "If you want to use log in with Facebook then Facebook will then be able to track your activity on this site. This can help them build a fuller picture of your online life.",
+		"description": ""
+	},
+	"inPageUI-tooltip-prompt-checkbox": {
+		"message": "Don't show me this again",
+		"description": ""
+	}
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply src/_locales/en/messages.json]: Reverted back and matched the white space!

File: src/content_script.js [hunk 11]
```diff
@@ -1,6 +1,54 @@
 "use strict";
 
-function detectFacebookLoginButton() {
+// Use the following patterns to check for on-screen Facebook elements
+
+const PATTERN_DETECTION_SELECTORS = [
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply src/content_script.js]: No – I had left the original mixed-group of patterns in the code. I'll remove it.

File: src/content_script.js [hunk 12]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
-  "[href*='facebook.com/dialog/share']",
-  "[data-bfa-network*='facebook']",
-  "[data-oauthserver*='facebook']", // Stackoverflow
-  "[id*='facebook_connect_button']", // Quora
-  "[aria-label*='Facebook']",
-  "[aria-label*='share on facebook']", // MSN
-  "[data-destination*='facebook']"
-];
+const fragmentClasses = ["fbc-badge-fence", "fbc-badge-tooltip", "fbc-badge-prompt"];
+const htmlBadgeFragmentPromptParagraphStrings = ["Allow Facebook to track you here?", "If you want to use log in with Facebook then Facebook will then be able to track your activity on this site. This can help them build a fuller picture of your online life."];
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply src/content_script.js]: Revised. They now pull from i18n.

File: src/content_script.js [hunk 13]
```diff
@@ -32,50 +80,411 @@ function itemWidthCheck (target) {
...
+  }
+  return false;
+}
+
+function addFacebookBadge (target, badgeClassUId, socialAction) {
+  // Detect if target is visible
+
+  const htmlBadgeDiv = createBadgeFragment(socialAction);
+
+  const htmlBadgeFragmentPromptButtonCancel = htmlBadgeDiv.querySelector(".fbc-badge-prompt-button-0");
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply src/content_script.js]: Resolved. New classes now read `fbc-badge-prompt-btn-cancel` and `fbc-badge-prompt-btn-allow`. The `btn-allow`/`cancel` match the i18n identifier as well.

File: src/content_script.js (hunk 2 above)
Comment (maxxcrawford) [reply src/content_script.js]: Understood. Revised!

File: src/content_script.js (hunk 3 above)
Comment (maxxcrawford) [reply src/content_script.js]: Revised!

**Comment by maxxcrawford** [comment]
@groovecoder @stoically Moved open review issues to #296. Merging this per @groovecoder request as is. Thanks!

**Pull request merged**